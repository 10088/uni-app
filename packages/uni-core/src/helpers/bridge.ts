import { extend } from '@vue/shared'
// TODO 等待 vue3 的兼容模式自带emitter
import E from './TinyEmitter'

export function initBridge(
  namespace: 'service' | 'view'
): Partial<UniApp.UniServiceJSBridge> {
  // TODO vue3 compatibility builds
  const emitter = new E()
  return extend(emitter, {
    subscribe(event: string, callback: Function): void {
      emitter.on(`${namespace}.${event}`, callback)
    },
    unsubscribe(event: string, callback: Function): void {
      emitter.off(`${namespace}.${event}`, callback)
    },
    subscribeHandler(event: string, args: unknown, pageId: number): void {
      if (__DEV__) {
        console.log(
          `[${namespace}][subscribeHandler][${Date.now()}]:${event}, ${JSON.stringify(
            args
          )}, ${pageId}`
        )
      }
      emitter.emit(`${namespace}.${event}`, args, pageId)
    },
  })
}
