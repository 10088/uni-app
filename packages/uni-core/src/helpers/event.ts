import { withModifiers } from 'vue'
export const onTouchmovePrevent = /*#__PURE__*/ withModifiers(() => {}, [
  'prevent',
])
export const onTouchmoveStop = /*#__PURE__*/ withModifiers(() => {}, ['stop'])
