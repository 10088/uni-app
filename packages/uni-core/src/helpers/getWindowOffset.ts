import safeAreaInsets from 'safe-area-insets'

export function getWindowOffset() {
  const style = document.documentElement.style
  const top = parseInt(style.getPropertyValue('--window-top'))
  const bottom = parseInt(style.getPropertyValue('--window-bottom'))
  const left = parseInt(style.getPropertyValue('--window-left'))
  const right = parseInt(style.getPropertyValue('--window-right'))
  return {
    top: top ? top + safeAreaInsets.top : 0,
    bottom: bottom ? bottom + safeAreaInsets.bottom : 0,
    left: left ? left + safeAreaInsets.left : 0,
    right: right ? right + safeAreaInsets.right : 0,
  }
}
