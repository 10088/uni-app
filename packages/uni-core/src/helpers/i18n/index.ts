import { initVueI18n } from '@dcloudio/uni-i18n'

import en from './en.json'
import es from './es.json'
import fr from './fr.json'
import zhHans from './zh-Hans.json'
import zhHant from './zh-Hant.json'

const messages = {
  en,
  es,
  fr,
  'zh-Hans': zhHans,
  'zh-Hant': zhHant,
}

const fallbackLocale = 'en'

const i18n = /*#__PURE__*/ initVueI18n(
  __PLATFORM__ === 'app-plus' || __PLATFORM__ === 'h5' ? messages : {},
  fallbackLocale
)

export function useI18n() {
  return i18n
}
