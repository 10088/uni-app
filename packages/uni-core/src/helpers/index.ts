export * from './util'
export * from './icon'
export * from './scroll'
export * from './getRealRoute'
export * from './updateCssVar'
export * from './getWindowOffset'
