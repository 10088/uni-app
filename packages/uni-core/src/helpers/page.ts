import { ComponentPublicInstance, getCurrentInstance } from 'vue'

export function useCurrentPageId() {
  return getCurrentInstance()!.root.proxy!.$page.id
}

export function getPageIdByVm(vm: ComponentPublicInstance) {
  if (!vm.$) {
    return
  }
  const rootProxy = vm.$.root.proxy
  if (rootProxy && rootProxy.$page) {
    return rootProxy.$page.id
  }
}
