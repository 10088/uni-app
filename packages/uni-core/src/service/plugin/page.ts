import { isString } from '@vue/shared'
import { ComponentPublicInstance } from 'vue'
import { invokeArrayFns } from '@dcloudio/uni-shared'

export function getCurrentPage() {
  const pages = getCurrentPages()
  const len = pages.length
  if (len) {
    return pages[len - 1]
  }
}

export function getCurrentPageMeta() {
  const page = getCurrentPage()
  if (page) {
    return page.$page.meta
  }
}

export function getCurrentPageId() {
  const meta = getCurrentPageMeta()
  if (meta) {
    return meta.id
  }
  return -1
}

export function getCurrentPageVm() {
  const page = getCurrentPage()
  if (page) {
    return (page as any).$vm as ComponentPublicInstance
  }
}

export function invokeHook(name: string, args?: unknown): unknown
export function invokeHook(id: number, name: string, args?: unknown): unknown
export function invokeHook(
  vm: ComponentPublicInstance,
  name: string,
  args?: unknown
): unknown
export function invokeHook(
  vm: ComponentPublicInstance | string | number,
  name?: string | unknown,
  args?: unknown
) {
  if (isString(vm)) {
    args = name
    name = vm
    vm = getCurrentPageVm()!
  } else if (typeof vm === 'number') {
    const page = getCurrentPages().find((page) => page.$page.id === vm)
    if (page) {
      vm = (page as any).$vm as ComponentPublicInstance
    } else {
      vm = getCurrentPageVm() as ComponentPublicInstance
    }
  }
  if (!vm) {
    return
  }
  const hooks = vm.$[name as string]
  return hooks && invokeArrayFns(hooks, args)
}
