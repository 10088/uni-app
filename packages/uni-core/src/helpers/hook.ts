import { ComponentPublicInstance } from 'vue'
import { isString } from '@vue/shared'
import { invokeArrayFns } from '@dcloudio/uni-shared'
import { getCurrentPageVm } from './page'

export function invokeHook(name: string, args?: unknown): unknown
export function invokeHook(id: number, name: string, args?: unknown): unknown
export function invokeHook(
  vm: ComponentPublicInstance,
  name: string,
  args?: unknown
): unknown
export function invokeHook(
  vm: ComponentPublicInstance | string | number,
  name?: string | unknown,
  args?: unknown
) {
  if (isString(vm)) {
    args = name
    name = vm
    vm = getCurrentPageVm()!
  } else if (typeof vm === 'number') {
    const page = getCurrentPages().find((page) => page.$page.id === vm)
    if (page) {
      vm = (page as any).$vm as ComponentPublicInstance
    } else {
      vm = getCurrentPageVm() as ComponentPublicInstance
    }
  }
  if (!vm) {
    return
  }
  const hooks = vm.$[name as string]
  return hooks && invokeArrayFns(hooks, args)
}

export function hasHook(vm: ComponentPublicInstance | number, name: string) {
  if (typeof vm === 'number') {
    const page = getCurrentPages().find((page) => page.$page.id === vm)
    if (page) {
      vm = (page as any).$vm as ComponentPublicInstance
    } else {
      vm = getCurrentPageVm() as ComponentPublicInstance
    }
  }
  if (!vm) {
    return false
  }
  const hooks = vm.$[name]
  return !!(hooks && hooks.length)
}
