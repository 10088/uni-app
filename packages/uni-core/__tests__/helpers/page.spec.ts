import { mergePageMeta } from '../../src/helpers/page'

function initDefaultUniConfig() {
  return JSON.parse(
    JSON.stringify({
      globalStyle: {
        navigationBar: {},
      },
    })
  )
}

describe('page', () => {
  test('mergePageMeta', () => {
    global.__uniConfig = initDefaultUniConfig()
    global.__uniConfig.globalStyle.navigationBar.titleText = 'uni-app'
    expect(
      mergePageMeta(1, {
        route: '',
        navigationBar: {},
      })
    ).toMatchSnapshot()
    expect(
      mergePageMeta(1, {
        route: '',
        navigationBar: { titleText: 'hello', titleColor: '#000000' },
        enablePullDownRefresh: true,
      })
    ).toMatchSnapshot()
    expect(
      mergePageMeta(1, {
        route: '',
        navigationBar: { titleColor: '#000000' },
        enablePullDownRefresh: true,
        pullToRefresh: {
          offset: 100,
        },
      })
    ).toMatchSnapshot()
  })
})
