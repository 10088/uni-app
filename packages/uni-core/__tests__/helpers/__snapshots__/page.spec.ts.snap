// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page mergePageMeta 1`] = `
Object {
  "id": 1,
  "navigationBar": Object {
    "titleText": "uni-app",
  },
  "pullToRefresh": Object {},
  "route": "",
}
`;

exports[`page mergePageMeta 2`] = `
Object {
  "enablePullDownRefresh": true,
  "id": 1,
  "navigationBar": Object {
    "titleColor": "#000000",
    "titleText": "hello",
  },
  "pullToRefresh": Object {},
  "route": "",
}
`;

exports[`page mergePageMeta 3`] = `
Object {
  "enablePullDownRefresh": true,
  "id": 1,
  "navigationBar": Object {
    "titleColor": "#000000",
    "titleText": "uni-app",
  },
  "pullToRefresh": Object {
    "offset": 100,
  },
  "route": "",
}
`;
